    Gallo, Florentina.
    Grillo, Gian Franco.
    Pizzingrilli, Federico.
    Quiroga, Gerardo.

# 

Imports para los ejercicios

In [7]:
import pandas as pd
import numpy as np
from sklearn.datasets import load_breast_cancer

In [18]:
#Seteo para ver todas las columnas cuando muestro un dataframe
pd.set_option('display.max_columns', None)

# 

# CONSIGNA

Presentar (máximo) tres sets de datos que puedan ser usados como fuente de información para sus proyectos finales. 

#  

# DATASETS

Los set de datos de interés para usar como fuente de información en el proyecto final son:


    1- The Spotify Hit Predictor Dataset (1960-2019).
    2- Breast Cancer Wisconsin (Diagnostic) Data Set.


Siendo de mayor interés utilizar el Dataset N°1

# 

## 1- The Spotify Hit Predictor Dataset (1960-2019).

In [ ]:
df_w = pd.read_csv('../../Datasets/dataset-of-10s.csv', sep=',')
df_w.info()

In [19]:
data = load_breast_cancer()

# Read the DataFrame, first using the feature data
df = pd.DataFrame(data.data, columns=data.feature_names)
# Add a target column, and fill it with the target data
df['target'] = data.target
# Show the first five rows
df


mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80    1001.00             0.12   
1          20.57         17.77          132.90    1326.00             0.08   
2          19.69         21.25          130.00    1203.00             0.11   
3          11.42         20.38           77.58     386.10             0.14   
4          20.29         14.34          135.10    1297.00             0.10   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00    1479.00             0.11   
565        20.13         28.25          131.20    1261.00             0.10   
566        16.60         28.08          108.30     858.10             0.08   
567        20.60         29.33          140.10    1265.00             0.12   
568         7.76         24.54           47.92     181.00             0.05   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0                0.28            0.30                 0.15           0.24   
1                0.08            0.09                 0.07           0.18   
2                0.16            0.20                 0.13           0.21   
3                0.28            0.24                 0.11           0.26   
4                0.13            0.20                 0.10           0.18   
..                ...             ...                  ...            ...   
564              0.12            0.24                 0.14           0.17   
565              0.10            0.14                 0.10           0.18   
566              0.10            0.09                 0.05           0.16   
567              0.28            0.35                 0.15           0.24   
568              0.04            0.00                 0.00           0.16   

     mean fractal dimension  radius error  texture error  perimeter error  \
0                      0.08          1.09           0.91             8.59   
1                      0.06          0.54           0.73             3.40   
2                      0.06          0.75           0.79             4.58   
3                      0.10          0.50           1.16             3.44   
4                      0.06          0.76           0.78             5.44   
..                      ...           ...            ...              ...   
564                    0.06          1.18           1.26             7.67   
565                    0.06          0.77           2.46             5.20   
566                    0.06          0.46           1.07             3.42   
567                    0.07          0.73           1.59             5.77   
568                    0.06          0.39           1.43             2.55   

     area error  smoothness error  compactness error  concavity error  \
0        153.40              0.01               0.05             0.05   
1         74.08              0.01               0.01             0.02   
2         94.03              0.01               0.04             0.04   
3         27.23              0.01               0.07             0.06   
4         94.44              0.01               0.02             0.06   
..          ...               ...                ...              ...   
564      158.70              0.01               0.03             0.05   
565       99.04              0.01               0.02             0.04   
566       48.55              0.01               0.04             0.05   
567       86.22              0.01               0.06             0.07   
568       19.15              0.01               0.00             0.00   

     concave points error  symmetry error  fractal dimension error  \
0                    0.02            0.03                     0.01   
1                    0.01            0.01                     0.00   
2                    0.02            0.02                     0.00   
3                    0.02            0.06                    

In [20]:
df_w = pd.read_csv('../../Datasets/dataset-of-10s.csv', sep=',')
df_w.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6398 entries, 0 to 6397
Data columns (total 19 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   track             6398 non-null   object 
 1   artist            6398 non-null   object 
 2   uri               6398 non-null   object 
 3   danceability      6398 non-null   float64
 4   energy            6398 non-null   float64
 5   key               6398 non-null   int64  
 6   loudness          6398 non-null   float64
 7   mode              6398 non-null   int64  
 8   speechiness       6398 non-null   float64
 9   acousticness      6398 non-null   float64
 10  instrumentalness  6398 non-null   float64
 11  liveness          6398 non-null   float64
 12  valence           6398 non-null   float64
 13  tempo             6398 non-null   float64
 14  duration_ms       6398 non-null   int64  
 15  time_signature    6398 non-null   int64  
 16  chorus_hit        6398 non-null   float64


# 

# 1

Cuantos ID tienen en común. Nombrar cuales no tienen en común

In [6]:
#Me quedo sólo con la columna ID de ambos df, y hago un inner join para quedarme sólo con ids que están en ambos df
#Luego, cuento los valores únicos

df_inner = df_w[['ID']].merge(df_p['ID'], how='inner', on='ID')

print("En total, tienen {:d} ID's en común".format(df_inner['ID'].nunique()))

En total, tienen 48842 ID's en común


In [7]:
#Hago un outer join para quedarme con los ids compartidos y los que no se comparten, pero esta vez, agregandole
#el parametro indicator=True, que me indica en que df está el valor por el que se joinea (ID).
#De estos, me quedo con todos los que no sean "both", ya que representa que esta en ambos df.
#Imprimo para mostrar los ID'S que no están compartidos en ambos df.

df_outer =  df_w[['ID']].merge(df_p['ID'], how='outer', on='ID', indicator=True)

df_outer = df_outer[df_outer['_merge']!='both']

df_outer[['ID']]

ID
48842  1344995623

# 

# 2

Corregir los target a solo dos valores distintos ‘<=50k’ y ‘>50k’

In [8]:
#Primero vemos los diferentes valores que vienen para saber como modificarlos

df_joined['Target'].unique()

array([' <=50K', ' >50K', nan, ' <=50K.', ' >50K.'], dtype=object)

In [9]:
#Vemos que tenemos un nulo, asi que lo dropeamos.
#Luego vemos los signos. Si contiene el signo '<' en la cadena, le ponemos '<=50K', sino, '>50K'


df_joined.dropna(subset=['Target'], inplace=True)

df_joined['Target'] = df_joined['Target'].apply(lambda target: '<=50K' if ('<' in target) else '>50K')


In [10]:
#Ahora podemos ver que tenemos sólo 2 valores

df_joined['Target'].unique()

array(['<=50K', '>50K'], dtype=object)

# 

# 3

Calcular la proporción que hay entre personas con salarios >50k respecto del total para 
Entre personas de distinto sexo, Raza y Educacion. Por ejemplo cuál es el porcentaje de mujeres con un salario mayor a 50K y compararlo con el de los hombres. Lo mismo para personas de distintas razas y Educación

In [11]:
#Creamos un df agrupando por la columna correspondiente, para cada variable solicitada, y contando la cantidad
#de registros de las mismas (cuento ID).
#Renombro las columnas ID por personas totales.

df_aggr_sex = df_joined[['ID','Sex']].groupby('Sex').agg({'ID' : 'count'})

df_aggr_race = df_joined[['ID','Race']].groupby('Race').agg({'ID' : 'count'})

df_aggr_education = df_joined[['ID','Education']].groupby('Education').agg({'ID' : 'count'})


df_aggr_sex.rename(columns={'ID' : 'total_persons'}, inplace=True)
df_aggr_race.rename(columns={'ID' : 'total_persons'}, inplace=True)
df_aggr_education.rename(columns={'ID' : 'total_persons'}, inplace=True)

In [12]:
#A cada df le agrego, por cada agrupación, la cantidad de personas con salarios >50K.
#Utilizo la columna ID porque es sobre la que se hace el count.
#Para facilitar lectura, creo un df filtrado con target >50K.


df_greater_than_50K = df_joined[df_joined['Target']=='>50K']

df_aggr_sex['>50K'] = df_greater_than_50K[['ID','Sex']].groupby('Sex').agg({'ID' : 'count'})['ID']

df_aggr_race['>50K'] = df_greater_than_50K[['ID','Race']].groupby('Race').agg({'ID' : 'count'})['ID']

df_aggr_education['>50K'] = df_greater_than_50K[['ID','Education']].groupby('Education').agg({'ID' : 'count'})['ID']


In [13]:
#Agrego una columna donde me indique el porcentaje con target >50K sobre el total de personas en el dataset.
#Esto se leería, en el caso de la agrupación por sexo: 
#"El X% del total de personas son mujeres con target >50K, y el Y% del total de personas son hombres con target >50K".


df_aggr_sex['perc_greater_than_50K_over_total'] = df_aggr_sex['>50K'].apply(lambda qty : 
                                                                    '{:.2%}'.format(qty / df_aggr_sex['total_persons'].sum()))


df_aggr_race['perc_greater_than_50K_over_total'] = df_aggr_race['>50K'].apply(lambda qty : 
                                                                    '{:.2%}'.format(qty / df_aggr_race['total_persons'].sum()))


df_aggr_education['perc_greater_than_50K_over_total'] = df_aggr_education['>50K'].apply(lambda qty : 
                                                                    '{:.2%}'.format(qty / df_aggr_education['total_persons'].sum()))


In [14]:
#Agrego una columna donde me indique el porcentaje con target >50K sobre la cantidad de personas de la clase actual.
#Esto se leería, en el caso de la agrupación por sexo: 
#"El X% de mujeres tienen target >50K, y el Y% de hombres tienen target >50K".


df_aggr_sex['perc_greater_than_50K_of_class'] = (df_aggr_sex['>50K'] / df_aggr_sex['total_persons']) *100


df_aggr_race['perc_greater_than_50K_of_class'] = (df_aggr_race['>50K'] / df_aggr_race['total_persons']) *100


df_aggr_education['perc_greater_than_50K_of_class'] = (df_aggr_education['>50K'] / df_aggr_education['total_persons']) *100

# 

### VEMOS LOS RESULTADOS 

In [15]:
#Quiero ver la información del df de los sexos, ordenando los sexos, según la cantidad de personas del sexo con target
#>50K, de mayor a menor

df_aggr_sex.sort_values('perc_greater_than_50K_of_class', ascending=False)

total_persons  >50K perc_greater_than_50K_over_total  \
Sex                                                             
 Male            32649  9917                           20.30%   
 Female          16192  1769                            3.62%   

         perc_greater_than_50K_of_class  
Sex                                      
 Male                             30.37  
 Female                           10.93

Podemos ver que el 30% de los hombres tienen target >50K, mientras que las mujeres solamente el 10% tiene target >50K

# 

In [16]:
#Quiero ver la información del df de las razas, ordenando los razas, según la cantidad de personas de la raza con target
#>50K, de mayor a menor

df_aggr_race.sort_values('perc_greater_than_50K_of_class', ascending=False)

total_persons   >50K perc_greater_than_50K_over_total  \
Race                                                                         
 Asian-Pac-Islander           1519    409                            0.84%   
 White                       41761  10606                           21.72%   
 Other                         406     50                            0.10%   
 Black                        4685    566                            1.16%   
 Amer-Indian-Eskimo            470     55                            0.11%   

                     perc_greater_than_50K_of_class  
Race                                                 
 Asian-Pac-Islander                           26.93  
 White                                        25.40  
 Other                                        12.32  
 Black                                        12.08  
 Amer-Indian-Eskimo                           11.70

Podemos ver que las personas de raza Asian-Pac-Islander son las que, dentro de su raza, tienen mayor porcentaje de personas con target >50K con casi un 27%

En lo opuesto, las personas de raza Amer-Indian-Eskimo son las que, dentro de su raza, tienen menor porcentaje de personas con target >50K con un 11%

Sin embargo, si analizamos, por raza, el porcentaje de personas con target >50K respecto al TOTAL de personas, notamos que la raza White tiene el mayor porcentaje de personas con target >50K respecto al total de personas, con un 21%

# 

In [17]:
#Quiero ver la información del df de educación, ordenando los niveles de educación, según la cantidad de personas del
#nivel de educación con target >50K, de mayor a menor

df_aggr_education.sort_values('perc_greater_than_50K_of_class', ascending=False)

total_persons  >50K perc_greater_than_50K_over_total  \
Education                                                             
 Prof-school             834   617                            1.26%   
 Doctorate               594   431                            0.88%   
 Masters                2657  1459                            2.99%   
 Bachelors              8024  3312                            6.78%   
 Assoc-acdm             1601   413                            0.85%   
 Assoc-voc              2061   522                            1.07%   
 Some-college          10878  2063                            4.22%   
 HS-grad               15784  2503                            5.12%   
 12th                    657    48                            0.10%   
 7th-8th                 955    62                            0.13%   
 10th                   1389    87                            0.18%   
 9th                     756    41                            0.08%   
 5th-6th                 509    27                            0.06%   
 11th                   1812    92                            0.19%   
 1st-4th                 247     8                            0.02%   
 Preschool                83     1                            0.00%   

               perc_greater_than_50K_of_class  
Education                                      
 Prof-school                            73.98  
 Doctorate                              72.56  
 Masters                                54.91  
 Bachelors                              41.28  
 Assoc-acdm                             25.80  
 Assoc-voc                              25.33  
 Some-college                           18.96  
 HS-grad                                15.86  
 12th                                    7.31  
 7th-8th                                 6.49  
 10th                                    6.26  
 9th                                     5.42  
 5th-6th                                 5.30  
 11th                                    5.08  
 1st-4th                                 3.24  
 Preschool                               1.20

Podemos ver que las personas con educación Prof-school son las que, dentro de su nivel de educación, tienen mayor porcentaje de personas con target >50K, con casi un 74%.

En lo opuesto, las personas con educación Preschool son las que, dentro de su nivel de educación, tienen menor porcentaje de personas con target >50K, con un 1%.

Sin embargo, si analizamos, por nivel de educación, el porcentaje de personas con target >50K respecto al TOTAL de personas, notamos que el nivel de educación Masters tiene el mayor porcentaje de personas con target >50K respecto al total de personas, con un 3%.

# 